### PR TITLE
Rosetta/sebp 83 pay coin operation

### PR DESCRIPTION
**Pull request by nikos-kitmeridis**
## Description 

**WIP**
Adds support for coin transfers request from Coinbase, for coins other than SUI.
In order to have the coin type in the Transaction data, the `ptb.pay()` method is not an option.
Instead, we are building the transaction manually with a Move call to the `transfer::public_transfer`.
Right now it returns a `VMVerificationOrDeserializationError` which is under investigation.
See test: `test_custom_coin_transfer`

## Test plan 

Unit and Integration tests
---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-rosetta/tests/custom_coins/test_coin_utils.rs [hunk 1]
```diff
@@ -132,7 +132,7 @@ pub async fn init_package(
     keystore: &Keystore,
     sender: SuiAddress,
 ) -> Result<InitRet> {
-    let path = Path::new("tests/custom_coins/test_coin");
+    let path = Path::new("/Users/nikos/Documents/vscodeProjects/sui/crates/sui-rosetta/tests/custom_coins/test_coin");
```
Comment (manolisliolios) [inline crates/sui-rosetta/tests/custom_coins/test_coin_utils.rs:135]: careful not to commit this path 😄

Comment (nikos-kitmeridis) [reply crates/sui-rosetta/tests/custom_coins/test_coin_utils.rs:135]: omg thanks for that!

File: crates/sui-rosetta/src/construction.rs [hunk 2]
```diff
@@ -306,13 +335,18 @@ pub async fn metadata(
                 .internal_operation
                 .try_into_data(ConstructionMetadata {
                     sender,
-                    coins: vec![],
+                    coins: gas_coin
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/construction.rs:338]: Why is this needed here and before it wasn't for the dry run?

Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/construction.rs:338]: This has been added in an attempt to see if its addition would solve the `VMVerificationOrDeserializationError` error, which apparently it didn't. So probably in the final PR it will be removed.

File: crates/sui-rosetta/src/construction.rs [hunk 3]
```diff
@@ -247,12 +251,37 @@ pub async fn metadata(
...
+                .map(|coin| coin.object_ref())
+                .collect();
+            let obj_id = coin_objs.first().unwrap().0.to_hex_literal();
+            println!("OBJ ID {}", obj_id);
+            gas_coin = context
+                .client
+                .coin_read_api()
+                .select_coins(sender, None, 5_000_000_u128, vec![])
+                .await
+                .ok();
```
Comment (patrickkuo) [inline crates/sui-rosetta/src/construction.rs:281]: we need to do a dry run to get the correct gas amount

File: crates/sui-rosetta/src/construction.rs (hunk 2 above)
Comment (manolisliolios) [reply crates/sui-rosetta/src/construction.rs:338]: yeah that shouldn't be the case since it works for the other scenarios (and it's only for gas logic)

File: crates/sui-rosetta/src/types.rs [hunk 4]
```diff
@@ -947,6 +959,63 @@ impl InternalOperation {
...
+                let mut recipient_map: IndexMap<SuiAddress, Vec<usize>> = IndexMap::new();
+                let mut amt_args = Vec::with_capacity(recipients.len());
+                for (i, (recipient, amount)) in recipients.into_iter().zip(amounts).enumerate() {
+                    recipient_map.entry(recipient).or_default().push(i);
+                    amt_args.push(builder.pure(amount)?);
+                }
+                let Argument::Result(split_primary) =
+                    builder.command(Command::SplitCoins(coin_arg, amt_args))
+                else {
+                    panic!("self.command should always give a Argument::Result")
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/types.rs:996]: I think you'd want an error, not a panic?

File: crates/sui-rosetta/src/types.rs [hunk 5]
```diff
@@ -947,6 +959,63 @@ impl InternalOperation {
...
+                };
+                let type_arg = TypeTag::from_str(&metadata.coin_type.clone().unwrap()).unwrap();
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/types.rs:998]: isn't `&metadata.coin_type` already a TypeTag?

Comment (manolisliolios) [reply crates/sui-rosetta/src/types.rs:998]: nvm it isn't, it's a `StructTag`, but I think you can easily use `.into()`:
https://github.com/MystenLabs/sui/blob/32347c20c248f6691c6ab89ba53266e2f677ed9b/external-crates/move/crates/move-core-types/src/language_storage.rs#L368

Comment (manolisliolios) [inline crates/sui-rosetta/src/types.rs:998]: I think you can directly call "into()" into TypeTag for your StructTag
https://github.com/MystenLabs/sui/blob/32347c20c248f6691c6ab89ba53266e2f677ed9b/external-crates/move/crates/move-core-types/src/language_storage.rs#L368

File: crates/sui-rosetta/src/types.rs [hunk 6]
```diff
@@ -947,6 +959,63 @@ impl InternalOperation {
...
+                let _ts = type_arg.to_canonical_string(true);
+                for (recipient, split_secondaries) in recipient_map {
+                    let rec_arg = builder.pure(recipient).unwrap();
+                    split_secondaries
+                        .into_iter()
+                        .map(|j| Argument::NestedResult(split_primary, j as u16))
+                        .for_each(|coin| {
+                            _ = builder.command(Command::move_call(
```
[the remark this answers is not in the document]
Comment (manolisliolios) [reply crates/sui-rosetta/src/types.rs:1006]: I don't think you need to write the move-call as is. 
You could also `bulkTransfer` them if there are more than 1 coins to be sent per single address (to avoid adding non-needed commands!) (using `transfer_args`)

[the remark this answers is not in the document]
Comment (manolisliolios) [reply crates/sui-rosetta/src/types.rs:1006]: Which brings me to the next question, why can't you utilize the `builder.pay` overall, to avoid doing the splits yourself too?

File: crates/sui-rosetta/src/types.rs (hunk 4 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/types.rs:996]: yep that's jsut a copy from the pay method, the PR is draft will take care of these details once I make the flow work.

Comment (manolisliolios) [reply crates/sui-rosetta/src/types.rs:996]: The thing is that `pay` uses the native PTB transfers, while you're trying to port the transfer logic into a move call `0x2::transfer::public_transfer` (hence adding the requirement for type params too). 

The only difference between a SUI transfer would be that you cannot use "gas" to do splits, so I don't think that should be that way.

File: crates/sui-rosetta/src/types.rs (hunk 5 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/types.rs:998]: Not following you on this. `metadata.coin_type` is an `Option<String>`

Comment (manolisliolios) [reply crates/sui-rosetta/src/types.rs:998]: Whoops, you're right, gh's preview gave me the wrong type.

File: crates/sui-rosetta/src/types.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/types.rs:1006]: > if there are more than 1 coins to be sent per single address

this is why the recipient map is being used (copied from the `builder.pay()` function)
```
                // collect recipients in the case where they are non-unique in order
                // to minimize the number of transfers that must be performed
                let mut recipient_map: IndexMap<SuiAddress, Vec<usize>> = IndexMap::new();
```

[the remark this answers is not in the document]
Comment (manolisliolios) [reply crates/sui-rosetta/src/types.rs:1006]: resolved + aligned over a call!

File: crates/sui-rosetta/src/operations.rs [hunk 7]
```diff
@@ -414,10 +461,37 @@ impl Operations {
                 aggregated_recipients
                     .into_iter()
                     .map(|(recipient, amount)| {
-                        Operation::pay_sui(status, recipient, amount.into())
+                        currency = input_args.as_ref().and_then(|args| {
```
Comment (patrickkuo) [inline crates/sui-rosetta/src/operations.rs:464]: can we add some comment here describing why we do this?

File: crates/sui-rosetta/src/operations.rs [hunk 8]
```diff
@@ -213,23 +249,33 @@ impl Operations {
...
+            ),
+            _ => Self::new(vec![Operation::generic_op(status, sender, tx)]),
         })
     }
 
     fn parse_programmable_transaction(
         sender: SuiAddress,
         status: Option<OperationStatus>,
         pt: SuiProgrammableTransactionBlock,
+        input_args: Option<Vec<CallArg>>,
```
Comment (patrickkuo) [inline crates/sui-rosetta/src/operations.rs:278]: does the `SuiProgrammableTransactionBlock` contains the call arg we need already?

File: crates/sui-rosetta/src/types.rs [hunk 9]
```diff
@@ -947,6 +956,17 @@ impl InternalOperation {
...
             }
+            Self::PayCoin {
+                recipients,
+                amounts,
+                ..
+            } => {
+                let mut builder = ProgrammableTransactionBuilder::new();
+                builder.pay(metadata.objects.clone(), recipients, amounts)?;
+                let currency_str = serde_json::to_string(&metadata.currency.unwrap()).unwrap();
+                builder.pure(currency_str)?;
```
Comment (patrickkuo) [inline crates/sui-rosetta/src/types.rs:967]: lets add comment here.

File: crates/sui-rosetta/src/operations.rs (hunk 8 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/operations.rs:278]: Hey @patrickkuo the reason for these `input_args` needed here is because while the `inputs` vector from the `SuiProgrammableTransactionBlock` contains the entry for this extra pure parameter I've added as the bearer of the currency info, it does not actually contain it's value. It seems that this data are not available in memory, I guess it has to do to the fact that this input is not effectively being used into any transaction (see attached img 1).
The workaround was to parse the raw tx block's data and deserialise the input objects so I can get the value in the extra `input_args` param you highlighted (see attached img 2). 
I' ll try to express this with a comment in the code as per your review comment above.

Screenshot 2024-08-07 at 2 24 16 PM
Screenshot 2024-08-07 at 2 25 06 PM

File: crates/sui-rosetta/src/types.rs (hunk 9 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/types.rs:974]: Comment added. Thx!

File: crates/sui-rosetta/src/operations.rs [hunk 10]
```diff
@@ -135,6 +139,38 @@ impl Operations {
         })
     }
 
+    fn pay_coin_ops_to_internal(self) -> Result<InternalOperation, Error> {
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/operations.rs:142]: `internal_pay_coins` might be more clear naming?

File: crates/sui-rosetta/src/operations.rs [hunk 11]
```diff
@@ -135,6 +139,38 @@ impl Operations {
...
+                    if amount > u64::MAX as i128 {
+                        return Err(Error::InvalidInput(
+                            "Input amount exceed u64::MAX".to_string(),
+                        ));
+                    }
+                    amounts.push(amount as u64)
+                }
+            }
+        }
+        let sender = sender.ok_or_else(|| Error::MissingInput("Sender address".to_string()))?;
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/operations.rs:164]: is that the only way we can find the sender?

File: crates/sui-rosetta/src/operations.rs [hunk 12]
```diff
@@ -213,23 +249,33 @@ impl Operations {
...
+        let sender = *data.sender();
+        Self::from_transaction(data.transaction().clone(), sender, None, input_args)
+    }
+
     fn from_transaction(
         tx: SuiTransactionBlockKind,
         sender: SuiAddress,
         status: Option<OperationStatus>,
-    ) -> Result<Vec<Operation>, Error> {
+        input_args: Option<Vec<CallArg>>,
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/operations.rs:264]: do we need option? Could we pass an empty vec?

File: crates/sui-rosetta/src/operations.rs (hunk 8 above)
Comment (patrickkuo) [reply crates/sui-rosetta/src/operations.rs:278]: it's stored as JsonValue

File: crates/sui-rosetta/src/operations.rs (hunk 11 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/operations.rs:164]: yep, remember how the incoming request payload looks like
```
[{
            "operation_identifier":{"index":0},
            "type":"PayCoin",
            "account": { "address" : "0x..." },
            "amount" : {
                "value": "30000000",
                "currency": {
                    "symbol": "TEST_COIN",
                    "decimals": 6,
                    "metadata": {
                        "coin_type": "0x...",
                    }
                }
            },
        },
        {
            "operation_identifier":{"index":1},
            "type":"PayCoin",
            "account": { "address" : "0x..." },
            "amount" : {
                "value": "-30000000",
                "currency": {
                    "symbol": "TEST_COIN",
                    "decimals": 6,
                    "metadata": {
                        "coin_type": "0x...",
                    }
                }
            },
        }]
```
which is in alignment to the existing PaySui operation. I wouldn't like to make this work in a different way at this point but something to be improved for both operations in a later point, if possible though, because it will most likely imply changes on the caller's side too.

File: crates/sui-rosetta/src/operations.rs (hunk 8 above)
Comment (patrickkuo) [reply crates/sui-rosetta/src/operations.rs:278]: ya I tried locally and it works.

you can use this code to get the bytes from json value
```
         currency = pt.inputs.iter().last().and_then(|arg| {
                            if let SuiCallArg::Pure(value) = arg {
                                let bytes = value
                                    .value()
                                    .to_json_value()
                                    .as_array()?
                                    .clone()
                                    .into_iter()
                                    .map(|v| v.as_u64().map(|n| n as u8))
                                    .collect::<Option<Vec<u8>>>()?;
                                bcs::from_bytes::<String>(&bytes)
                                    .ok()
                                    .and_then(|bcs_str| serde_json::from_str(&bcs_str).ok())
                            } else {
                                None
                            }
                        });
```

Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/operations.rs:278]: Awesome, thank you very much. I will amend accordingly then as it will simplify the flow.

File: crates/sui-rosetta/src/construction.rs [hunk 13]
```diff
@@ -239,6 +238,12 @@ pub async fn metadata(
     let option = request.options.ok_or(Error::MissingMetadata)?;
     let budget = option.budget;
     let sender = option.internal_operation.sender();
+    let currency = match &option.internal_operation {
+        InternalOperation::PayCoin { currency, .. } => Some(currency.clone()),
+        _ => None,
+    };
+    let coin_type = currency.clone().map(|c| c.metadata.coin_type);
```
Comment (nikos-terzo) [inline crates/sui-rosetta/src/construction.rs:245]: Minor: Instead of cloning the whole `currency` struct, you can use `currency.as_ref().map(|c| c.metadata.coin_type.clone())`.

File: crates/sui-rosetta/src/construction.rs [hunk 14]
```diff
@@ -369,6 +389,7 @@ pub async fn metadata(
             total_coin_value,
             gas_price,
             budget,
+            currency: currency.clone(),
```
Comment (nikos-terzo) [inline crates/sui-rosetta/src/construction.rs:392]: No need to clone here either

File: crates/sui-rosetta/src/operations.rs (hunk 8 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/operations.rs]: Done, thank you!

File: crates/sui-rosetta/src/operations.rs (hunk 12 above)
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/operations.rs]: Was about to replace with an empty vec as per your suggestion but apparently we won't need this property at all. thnx!

File: crates/sui-rosetta/src/construction.rs [hunk 15]
```diff
@@ -253,6 +258,20 @@ pub async fn metadata(
             let amount = amounts.iter().sum::<u64>();
             (Some(amount), vec![])
         }
+        InternalOperation::PayCoin { amounts, .. } => {
+            let amount = amounts.iter().sum::<u64>();
+            let coin_objs: Vec<ObjectRef> = context
+                .client
+                .coin_read_api()
+                .select_coins(sender, coin_type.clone(), amount.into(), vec![])
```
[the remark this answers is not in the document]
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/construction.rs]: fixed, thnx!

**Review by patrickkuo — APPROVED** [review]
LGTM, thanks for the change!
